### PR TITLE
Feature/search 213 optimize levenshtein startswith

### PR DESCRIPTION
**Pull request by Dronplane**
### Scope & Purpose
With adding prefix to the LEVENSHTEIN_MATCH function we could optimize existing queries having "simulated" prefix with constructs like 
`LEVENSHTEIN_MATCH(d.foo, 'foobar', 2) AND STARTS_WITH(d.foo, 'foo')`
That could be converted just to 
`LEVENSHTEIN_MATCH(d.foo, 'bar', 2, false, 64, 'foo')`

This will significantly speed up the execution as there will be not additional terms disjunction for STARTS_WITH and LEVENSHTEIN_MATCH execution will also cover less term variants.

Also check could be made if LEVENSHTEIN_MATCH && STARTS_WITH is ever possible. If length of STARTS_WITH prefix is greater than max possible term accepted by Levenshtein automaton - conjunction is replaced by "empty" filter.

Fixed boost factor setting for STARTS_WITH function.

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

File: arangod/Aql/IResearchViewOptimizerRules.cpp [hunk 1]
```diff
@@ -77,6 +77,52 @@ inline IResearchViewStoredValues const& storedValues(arangodb::LogicalView const
...
+    auto current = condition.getMemberUnchecked(memberIdx);
+    TRI_ASSERT(current);
+    auto const numAndMembers = current->numMembers();
+    if (current->type == arangodb::aql::AstNodeType::NODE_TYPE_OPERATOR_NARY_AND &&
+        numAndMembers > 1) {
+      size_t movePoint = numAndMembers - 1;
+      do {
+        auto candidate = current->getMemberUnchecked(movePoint);
+        if (candidate->type != arangodb::aql::AstNodeType::NODE_TYPE_FCALL ||
+            arangodb::iresearch::getFuncName(*candidate) != STARTS_WITH) {
```
Comment (gnusi) [inline arangod/Aql/IResearchViewOptimizerRules.cpp:97]: I suggest to compare actual function pointers

File: arangod/Aql/IResearchViewOptimizerRules.cpp [hunk 2]
```diff
@@ -77,6 +77,52 @@ inline IResearchViewStoredValues const& storedValues(arangodb::LogicalView const
...
+          break;
+        }
+      } while ((--movePoint) != 0);
+      for (size_t andMemberIdx = 0; andMemberIdx < movePoint; ++andMemberIdx) {
+        auto andMember = current->getMemberUnchecked(andMemberIdx);
+        if (andMember->type == arangodb::aql::AstNodeType::NODE_TYPE_FCALL &&
+            arangodb::iresearch::getFuncName(*andMember) == STARTS_WITH) {
```
Comment (gnusi) [inline arangod/Aql/IResearchViewOptimizerRules.cpp:104]: I suggest to compare actual function pointers

File: arangod/Aql/IResearchViewOptimizerRules.cpp [hunk 3]
```diff
@@ -77,6 +77,52 @@ inline IResearchViewStoredValues const& storedValues(arangodb::LogicalView const
...
+          TEMPORARILY_UNLOCK_NODE(current);
+          auto tmp = current->getMemberUnchecked(movePoint);
+          current->changeMember(movePoint--, andMember);
+          current->changeMember(andMemberIdx, tmp);
+          while (movePoint > andMemberIdx &&
+                 current->getMemberUnchecked(movePoint)->type ==
+                     arangodb::aql::AstNodeType::NODE_TYPE_FCALL &&
+                 arangodb::iresearch::getFuncName(
```
Comment (gnusi) [inline arangod/Aql/IResearchViewOptimizerRules.cpp:112]: I suggest to compare actual function pointers

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 4]
```diff
@@ -3570,17 +3569,68 @@ Result fromFuncStartsWith(
...
+    if (!isMultiPrefix && !prefixes.empty() &&
+         ctx.allowFiltersMerge && filter->type() == irs::type<irs::And>::id()) {
+      for (auto& f : *filter) {
+        if (f.type() == irs::type<irs::by_edit_distance>::id()) {
+          auto& levenshtein = static_cast<irs::by_edit_distance&>(f);
+          if (*levenshtein.mutable_field() == name) {
+            auto options = levenshtein.mutable_options();
+            auto startsWith = prefixes.back().second;
+            if (startsWith.size() <= options->prefix.size()) {
+              if (std::memcmp(options->prefix.data(), startsWith.c_str(),
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:3583]: can also use `irs::starts_with` for simplicity

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 5]
```diff
@@ -3570,17 +3569,68 @@ Result fromFuncStartsWith(
...
+                              startsWith.size()) == 0) {
+                // Nothing to do. We are already covered by this levenshtein prefix
+                return {};
+              }
+            } else {
+              // maybe we could enlarge prefix to cover us?
+              if (std::memcmp(options->prefix.data(), startsWith.c_str(),
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:3590]: can also use irs::starts_with for simplicity

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 6]
```diff
@@ -3570,17 +3569,68 @@ Result fromFuncStartsWith(
...
+                              options->prefix.size()) == 0) {
+                // looks promising - beginning of the levenshtein prefix is ok
+                auto prefixTailSize = startsWith.size() - options->prefix.size();
+                if (options->term.size() >= prefixTailSize &&
+                    std::memcmp(options->term.data(),
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:3595]: can also use irs::starts_with for simplicity

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 7]
```diff
@@ -3570,17 +3569,68 @@ Result fromFuncStartsWith(
 
     TRI_ASSERT(filterCtx.analyzer);
     kludge::mangleField(name, filterCtx.analyzer);
-    filter->boost(filterCtx.boost);
+
+    // Try to optimize us away
+    if (!isMultiPrefix && !prefixes.empty() &&
+         ctx.allowFiltersMerge && filter->type() == irs::type<irs::And>::id()) {
+      for (auto& f : *filter) {
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:3576]: I'd opt for moving this optmization logic into a separate cpp unit so we can start collecting such optimization for further backporting to iresaearch

**Comment by Dronplane** [comment]
https://jenkins.arangodb.biz:3456/view/PR/job/arangodb-matrix-pr/16976/

**Comment by Dronplane** [comment]
https://jenkins.arangodb.biz:3456/view/PR/job/arangodb-matrix-pr/16977/ tests blue

**Review by gnusi — APPROVED** [review]

**Pull request merged**